### PR TITLE
Roll-your-own NUTS region lookup to replace postcode.io

### PR DESCRIPTION
**Pull request by gmackay**
Service now builds a static map of polygons based on JSON data then queries this map to return region from lat/long parameters

removed call to postcode.io

[CSHRCP-311]

### CODE REVIEW

File: location-service/src/test/java/uk/gov/cshr/locationservice/service/RegionLookupServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import org.junit.Assert;
+import org.testng.annotations.Test;
+
+public class RegionLookupServiceTest {
+
+
+    @Test
+    public void testRegionLookups() throws Exception {
+
+        Assert.assertEquals(UK_NUTS.UKC, RegionLookup.findRegion(55.2970314, -1.72889996));
```
Comment (seamussingh) [inline location-service/src/test/java/uk/gov/cshr/locationservice/service/RegionLookupServiceTest.java]: Please include a test for null coordinates supplied and coordinates that would not be found in any region

File: location-service/src/main/java/uk/gov/cshr/locationservice/UK_NUTS.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package uk.gov.cshr.locationservice;
```
Comment (seamussingh) [inline location-service/src/main/java/uk/gov/cshr/locationservice/UK_NUTS.java]: This class should be in a util package or something similar not a top level class at the same level as the entry point.

File: location-service/src/test/java/uk/gov/cshr/locationservice/service/RegionLookupServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package uk.gov.cshr.locationservice.service;
```
Comment (seamussingh) [inline location-service/src/test/java/uk/gov/cshr/locationservice/service/RegionLookupServiceTest.java]: This class should be in the same package as RegionLoookup

File: location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                        pathsList.add(polygon);
+                    }
+                }
+            }
+        }
+
+        return ukNutsMap;
+    }
+
+    public static UK_NUTS findRegion(Double latitude, Double longitude) throws RuntimeException {
```
[the remark this answers is not in the document]
Comment (reachsaurabh) [reply location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java]: agreed

File: location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+public class RegionLookup {
+
+    private static final Logger log = LoggerFactory.getLogger(RegionLookup.class);
+
+    private static HashMap<UK_NUTS, List<Path2D.Double>> nutsMap;
+
+    private RegionLookup() {
+    }
+
+    public static HashMap<UK_NUTS, List<Path2D.Double>> createNutsMap() throws IOException {
```
[the remark this answers is not in the document]
Comment (reachsaurabh) [reply location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java]: agreed

**Comment by reachsaurabh** [comment]
Thanks @seamussingh , most points you raised are valid , refactoring/cleaning up needed

File: location-service/src/test/java/uk/gov/cshr/locationservice/service/RegionLookupServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+import uk.gov.cshr.locationservice.RegionLookup;
+import uk.gov.cshr.locationservice.UK_NUTS;
+import org.junit.Assert;
+import org.testng.annotations.Test;
+
+public class RegionLookupServiceTest {
```
[the remark this answers is not in the document]
Comment (reachsaurabh) [reply location-service/src/test/java/uk/gov/cshr/locationservice/service/RegionLookupServiceTest.java]: agreed

File: location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RegionLookup {
```
[the remark this answers is not in the document]
Comment (gmackay) [reply location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java]: Wholeheartedly disagree.  The location service returns a Coordinate object containing a latitude, longitude and a region.  How it does that is contained within the service,

To further enforce this level of encapsulation, I'll move the RegionLookup class into the service package make it package protected.

File: location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+            String nutsID = featureNode.get("properties").get("NUTS_ID").asText();
+
+            UK_NUTS ukNut;
+
+            try {
+                ukNut = UK_NUTS.valueOf(nutsID);
+            }
+            catch (IllegalArgumentException e) {
+                log.debug("nutsID:" + nutsID + " not recognised");
```
[the remark this answers is not in the document]
Comment (gmackay) [reply location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java]: It's not an error.  We're only doing UK NUTS.

File: location-service/src/main/java/uk/gov/cshr/locationservice/UK_NUTS.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    UKG("West Midlands"),
+    UKH("East of England"),
+    UKI("Greater London"),
+    UKJ("South East"),
+    UKK("South West"),
+    UKL("Wales"),
+    UKM("Scotland"),
+    UKN("Northern Ireland");
+
+    final private String name;
```
[the remark this answers is not in the document]
Comment (gmackay) [reply location-service/src/main/java/uk/gov/cshr/locationservice/UK_NUTS.java]: Seriously ;)

File: location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+                    if (polygon.contains(latitude, longitude)) {
+                        return ukNuts;
+                    }
+                }
+            }
+
+            return null;
+        }
+        catch (IOException ioe) {
+            throw new RuntimeException("Exception initialising the NUTS map", ioe);
```
[the remark this answers is not in the document]
Comment (gmackay) [reply location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java]: I disagree.  The exception has been converted from a checked (IOException) to a runtime because it indicates something has gone wrong which is out of control of the program namely that the data file could not be read.

This is a setup issue, is non recoverable, and should be thrown up the stack.

Checked exceptions should be recoverable while runtime exceptions indicate the application is effectively broken.

"Here's the bottom line guideline: If a client can reasonably be expected to recover from an exception, make it a checked exception. If a client cannot do anything to recover from the exception, make it an unchecked exception."
https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html

File: location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (gmackay) [reply location-service/src/main/java/uk/gov/cshr/locationservice/RegionLookup.java]: This is private implementation and I think TheNomenclatureOfTerritorialUnitsForStatistics for an enum is unwieldy.

File: location-service/src/test/java/uk/gov/cshr/locationservice/service/RegionLookupServiceTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (gmackay) [reply location-service/src/test/java/uk/gov/cshr/locationservice/service/RegionLookupServiceTest.java]: The tests check the correct Enum value is returned.  I can't see the merit in testing the actual name property of each enum value.  This might change anyway if we decide to return the Enum rather than the string which might be better for language translation.

**Comment by gmackay** [comment]
Refactored/tidied up  code.
Quick bit of performance testing seems good.
